docs/libvirt-pod-networking.md

```diff
@@ -0,0 +1,98 @@
+# Using CNI Interfaces and libvirt for networking
+
+Author: Fabian Deutsch \<fabiand@redhat.com\>
+
+## Introduction
+
+Today networking in KubeVirt is not integrated with Kubernetes, a user can
+rather choose to create a specific VM Spec definition to create a particular
+libvirt domxml, to reuse an existing physical device.
+But this is obviously suboptimal
+
+- It is not integrated with Kubernetes
+- It does not provide connectivity with pods
+- It requires to specify host specific bits (the network device to use)
+
+The purpose of this document is to describe one approach of how we can provide
+a minimal integrated solution, with few workarounds.
+
+**Note:** This approach is rough, and not optimized, but the rule of thumb was
+to find a robust solution which wrks with Kubernetes today.
+
+
+## Use-case
+
+The primary use-case is to allow VM to get connected to the pod network.
+
+
+## API
+
+The API is pretty clear: As usual an interface will be defined.
+To indicate that Pod interfaces shall be used for networking, a new type
+`pod-network` is introduced.
+All other fields, concerning the properties of the virtual NIC, can be used
+as usual.
+
+```yaml
+kind: VM
+spec:
+  domain:
+    devices:
+      interfaces:
+      - type: pod-network
+        mac:
+          address: 00:11:22:33:44:55
+        model:
+          type: virtio
+        source:
+          network: default
+```
+
+The `source` field can be omitted initially, as Kubernetes only provides a
+single network for Pods today.
+In future we might want to allow it to select the Pod network to connect to.
+
+
+## Implementation
+
+Conceptually the implementation works by assuming that every vNIC will be
+mapped to a Pod interface. To achieve this, for every vNIC a _new_ Pod
+interface is requested through CNI.
+
+**Note:** The original Pod interface (eth0) is never touched and can be used to
+access libvirtd (thorugh the libvirtd Pod IP) or to provide VM centric services
+through the VM Pod IP.
+
+**Note:** The _new_ Pod interfaces can not be directly attached to the VM,
+because we loose the ability to provide the IP to the guest. The IP can only be
+provided by DHCP, which requires the use of a bridge.
+
+The broader steps to enable networking in the described way:
+
+1. Request/Create a _new_ Pod interface on the libvirt pod
+2. Remember the IP of the Pod interface, then remove it from the Pod interface
+3. Create a libvirt network, backed by a bridge, enable DHCP,
+   and add a single host entry, with the IP address of the Pod, and the MAC
+   address of the corresponding virtual NIC of the VM
```

Comment:
The scalability of the libvirt network driver is unknown, because I don't believe any application has ever used it to manage more than a handful of networks at a time.  If there are scalability problems in the libvirtd code related to networks, those should be relatively easy to get addressed. We can't do much about the need to have many dnsmasq instances running - one for each network - that could turn into a noticeable memory drain, which is bad because memory is already the most constrained resource on typical compute nodes.